### PR TITLE
Return proper error codes in machines controller

### PR DESCRIPTION
**Pull request by rsalevsky**

### CODE REVIEW

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 1]
```diff
@@ -71,39 +71,30 @@ def show
...
     @machine.intended_role = params[:role]
-    @machine.save
 
     respond_to do |format|
-      format.json { head :ok }
+      if @machine.save
+        format.json { head :ok }
+      else
+        format.json do
+          render json: { error: "Cannot save node #{@machine.name}" }, status: 422
```
Comment (tboerger) [inline crowbar_framework/app/controllers/machines_controller.rb]: Please use ```status: :unprocessable_entity```

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 2]
```diff
@@ -71,39 +71,30 @@ def show
...
-    raise "Not allowed for admin nodes" if @machine.admin?
-    @machine.delete
 
     respond_to do |format|
-      format.json { head :ok }
+      if @machine.save
+        format.json { head :ok }
+      else
+        format.json do
+          render json: { error: "Cannot save node #{@machine.name}" }, status: 422
```
Comment (tboerger) [inline crowbar_framework/app/controllers/machines_controller.rb]: Please use ```status: :unprocessable_entity```

**Comment by tboerger** [comment]
In general it looks already ok, just some NITPIK... Please format the json blocks in a more generic way and don't mix ```{ }``` with ```do end```. Can you create a pull request for the node subcommand on https://github.com/crowbar/crowbar-client?

**Comment by tboerger** [comment]
Oh, tests failed...

**Comment by MaximilianMeister** [comment]
yeah need to adapt `machines_controller_spec.rb`

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 3]
```diff
@@ -57,7 +57,7 @@ def index
     end
 
     respond_to do |format|
-      format.json { render :json => { :nodes => @nodes }, :status => 200 }
+      format.json { render :json => { :nodes => @nodes }, :status => :ok }
```
Comment (tboerger) [inline crowbar_framework/app/controllers/machines_controller.rb]: We can skip this status, 200 is the default

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 4]
```diff
@@ -71,39 +71,30 @@ def show
...
     @machine.intended_role = params[:role]
-    @machine.save
 
     respond_to do |format|
-      format.json { head :ok }
+      if @machine.save
+        format.json { head :ok }
+      else
+        format.json do
+          render json: { error: "Cannot save node #{@machine.name}" }, status: :unprocessable_entity
```
Comment (tboerger) [inline crowbar_framework/app/controllers/machines_controller.rb]: I18n?

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 5]
```diff
@@ -71,39 +71,30 @@ def show
...
-    raise "Not allowed for admin nodes" if @machine.admin?
-    @machine.delete
 
     respond_to do |format|
-      format.json { head :ok }
+      if @machine.save
+        format.json { head :ok }
+      else
+        format.json do
+          render json: { error: "Cannot save node #{@machine.name}" }, status: :unprocessable_entity
```
Comment (tboerger) [inline crowbar_framework/app/controllers/machines_controller.rb]: I18n?

File: crowbar_framework/spec/controllers/machines_controller_spec.rb [hunk 6]
```diff
@@ -101,61 +99,74 @@
...
+    context "for existent node" do
+      it "assignes role compute" do
+        NodeObject.any_instance.expects(:intended_role=).with("compute").once
+
+        post :role, :name => "testing", :role => "compute", :format => "json"
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    it "returns not_found (404) when for non-existent node" do
```
Comment (tboerger) [inline crowbar_framework/spec/controllers/machines_controller_spec.rb]: Something like return 404 HTTP status

File: crowbar_framework/spec/controllers/machines_controller_spec.rb [hunk 7]
```diff
@@ -101,61 +99,74 @@
...
+      post :role, :name => "nonexistent", :format => "json"
+      expect(response).to have_http_status(:not_found)
+    end
+
+    it "returns unprocessable_entity (422) when save fails" do
```
Comment (tboerger) [inline crowbar_framework/spec/controllers/machines_controller_spec.rb]: Something like return 422 HTTP status

File: crowbar_framework/spec/controllers/machines_controller_spec.rb [hunk 8]
```diff
@@ -101,61 +99,74 @@
...
+        expect(response).to have_http_status(:not_found)
       end
     end
-  end
 
-  describe "DELETE delete" do
-    context "for existent node" do
-      it "invokes delete" do
-        NodeObject.any_instance.expects(:delete).once
+    it "returns unprocessable_entity (422) when save fails" do
```
Comment (tboerger) [inline crowbar_framework/spec/controllers/machines_controller_spec.rb]: Same as above

**Comment by tboerger** [comment]
Some NITPIK but :+1: beside that

**Comment by tboerger** [comment]
:+1:

File: crowbar_framework/app/controllers/machines_controller.rb [hunk 9]
```diff
@@ -116,15 +109,25 @@ def delete
...
     :powercycle,
     :poweroff,
-    :allocate
+    :allocate,
+    :delete,
+    :identify
   ].each do |action|
     add_help(action, [:id], [:post])
     define_method action do
       raise "Not allowed for admin nodes" if @machine.admin?
```
Comment (vuntz) [inline crowbar_framework/app/controllers/machines_controller.rb]: This is actually not correct for `identify`.

File: crowbar_framework/app/models/node_object.rb [hunk 10]
```diff
@@ -1172,17 +1181,19 @@ def bmc_cmd(cmd)
         ssh_command = "/sbin/poweroff -f"
       else
         Rails.logger.warn("ipmitool #{cmd} failed for #{@node.name}.")
-        return nil
+        return [422, "ipmitool #{cmd} failed for #{@node.name}"]
```
Comment (vuntz) [inline crowbar_framework/app/models/node_object.rb]: No translation here?

File: crowbar_framework/app/models/node_object.rb [hunk 11]
```diff
@@ -1143,21 +1147,26 @@ def net_rpc_cmd(cmd)
...
+        [422, "Node in unknown state"]
       end
     when :shutdown
       unless system("net", "rpc", "shutdown", "-I", @node.name ,"-U", "Administrator%#{@node[:provisioner][:windows][:admin_password]}")
         Rails.logger.warn("samba command \"#{cmd}\" for #{@node.name} failed - node in unknown state")
+        [422, "Node in unknown state"]
       end
     else
       Rails.logger.warn("samba command #{cmd} failed for #{@node.name}.")
+      [422, "Command #{cmd} failed"]
```
Comment (vuntz) [inline crowbar_framework/app/models/node_object.rb]: Actually, the log was wrong and your text is therefore wrong. It's more like "unknown command".

**Comment by MaximilianMeister** [comment]
i have only git related things to add here, the rest is good IMO

I would squash the i18n messages with the return proper error codes commit https://github.com/rsalevsky/barclamp-crowbar/commit/0847c4bebbf4cfa3ca7263c6468a2baeb586cc36 as it belongs together IMO

and in general I would squash as well the machines_controller commits with the spec that belongs to it like you did in https://github.com/rsalevsky/barclamp-crowbar/commit/e872019fbd1a5835fc06e29c648f0c1bcc8e8a75

**Comment by MaximilianMeister** [comment]
+1

**Comment by tboerger** [comment]
Some useful squashing than you have my :+1:

**Comment by rsalevsky** [comment]
@tboerger Squashed some commits, should be fine now and ready for merging. All tests are green...

**Comment by tboerger** [comment]
:+1:

**Pull request merged**